### DISCUSSION TITLE
[DISCUSSION] Why QEMU is required

### DISCUSSION THREAD

**Discussion by lucasew** [nix-community/nix-on-droid#143] [General]
I just want to understand

Why qemu-user is necessary?

What causes the overhead that makes termux often more than 3x faster on syscall heavy stuff?

BTW, I was experimenting with proot here without problems without relying on QEMU and even tested running proot via ADB to run an alpine chroot on an old galaxy pocket without success because for some reason I was unable to override the temporary directory xD.

When I have more free time I would like to hack around in this repo to find a way to make it faster but first I have to understand the more under the hood stuff xD.

**Comment by t184256** [comment]
> Why qemu-user is necessary?

qemu-user should be necessary for a split second at most to initialize nix store during the bootstrap zipball building. This happens on my (or zipball rebuilder's) side; nothing that's executed on-device needs qemu-user.

> What causes the overhead that makes termux often more than 3x faster on syscall heavy stuff?

proot does. proot traps syscalls using ptrace and fakes familiar locations for `/nix/store`, `/etc` and other select directories. that incurs some serious runtime overhead.

> When I have more free time I would like to hack around in this repo to find a way to make it faster

The straightest way would've been to stop using it, but that doesn't just mean patching everything that wanted to access /etc for a new location, that also means rebuilding nixpkgs with a custom store path. I don't have this much aarch64 power spare, or even non-aarch64 power to spare.
